### PR TITLE
add authorized fun for user

### LINKED ISSUE #15: auth middleware for user

### PR DESCRIPTION
**Pull request by Nader-SH**

### CODE REVIEW

File: server/middlewares/auth.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { Request, Response, NextFunction } from 'express';
+
+require('dotenv').config();
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: No need to require it here, you already required It in the server.

File: server/middlewares/auth.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const jwt = require('jsonwebtoken');
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: Use import

File: server/middlewares/auth.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+const verifyAccessToken = (req: Request, res: Response, next: NextFunction) => {
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: change the name to `checkAuth`

File: server/middlewares/auth.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        phone:number,
+        location:string,
+        img:string,
+        description:string,
+        password:string,
+        license_number:number,
+        status:string,
+        owner_name:string,
+        owner_img:string,
+        owner_id:number
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: We just need the id and the role

File: server/middlewares/auth.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    const { token } = req.cookies;
+    if (!token) throw new Error('Unauthorized');
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: We need to create a Custom Error to get our message and status

File: server/middlewares/auth.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    try {
+      jwt.verify(token, process.env.SECRET_KEY, (err: Error, decoded: UserDataType) => {
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: Convert this function to Promise in another folder and use it here with await

File: server/middlewares/auth.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        if (err) {
+          throw new Error('Unauthorized');
+        } else {
+          req.user = decoded;
+          next();
+        }
+      });
+    } catch (err) {
+      res.status(401);
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: Why is the status here? you don't send a response

File: server/middlewares/auth.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      throw new Error('Unauthorized');
```
Comment (Raghad-Mezied) [inline server/middlewares/auth.ts]: You need to pass the error here to the error middleware, so we can send a general response there.

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

**Review by Ibtisam-Hemmo — APPROVED** [review]

**Review by AhmedAbuSharar — APPROVED** [review]

**Pull request merged**